[css-forms-1] Add new pseudo-elements for customizable select

josepharhar opened the pull request:
This PR adds the ::check and ::select-arrow pseudo-elements so that select elements can have a standardized and customizable dropdown icon on the button and checkmarks next to options.

Fixes https://github.com/w3c/csswg-drafts/issues/10908

josepharhar commented:
I mostly copied the definitions from before/after. If before/after aren't actually part-like, then should I copy the text from ::backdrop that looks like this instead of saying these pseudos are part like?

> The ::backdrop pseudo-element is a tree-abiding pseudo-element. It accepts all properties. It inherits from its originating element.

dbaron commented on css-pseudo-4/Overview.bs:
```diff
@@ -1634,6 +1634,32 @@ Styling form control pickers: the ''::picker()'' pseudo-element</h3>
+  restriction on which properties apply to the ''::select-arrow'' pseudo-element.
+
+  ''::select-arrow'' generates a box as if it was an immediate child of its
+  <a>originating element</a>, preceding any boxes generated by the ''::before''
```
So I'm wondering if the idea that it precedes the `::before` is incompatible with being part-like.  I'm a little worried that it might be.

josepharhar replied:
I made it not part like anymore. Does that resolve this?

dbaron commented on css-pseudo-4/Overview.bs:
```diff
@@ -1634,6 +1634,32 @@ Styling form control pickers: the ''::picker()'' pseudo-element</h3>
+
+  ''::select-arrow'' generates a box as if it was an immediate child of its
+  <a>originating element</a>, preceding any boxes generated by the ''::before''
+  pseudo-element, with content as specified by 'content'.
```
I'm also worried about allowing `content` here when it doesn't (in reality) work on elements, which part-like pseudo-elements should fully behave like.

josepharhar replied:
I made it not part like anymore. Does that resolve this?

dbaron commented on css-pseudo-4/Overview.bs:
```diff
@@ -1634,6 +1634,32 @@ Styling form control pickers: the ''::picker()'' pseudo-element</h3>
+  ''::select-arrow'' is a [=part-like pseudo-element=], so there is no
+  restriction on which properties apply to the ''::select-arrow'' pseudo-element.
+
+  ''::select-arrow'' generates a box as if it was an immediate child of its
```
I don't like the term "immediate child"; "child" is fine.  You don't need "immediate" to differentiate "child" from "descendant".

josepharhar replied:
I agree, done

dbaron commented on css-pseudo-4/Overview.bs:
```diff
@@ -1634,6 +1634,32 @@ Styling form control pickers: the ''::picker()'' pseudo-element</h3>
+  ''::select-arrow'' is a [=part-like pseudo-element=], so there is no
+  restriction on which properties apply to the ''::select-arrow'' pseudo-element.
+
+  ''::select-arrow'' generates a box as if it was an immediate child of its
```
Are there conditions on when this generates a box?  Does it only do something for `appearance: base-select`?

josepharhar replied:
Sure, I added the restriction that there must be a select element with base appearance.

dbaron reviewed (commented):
A few comments (and questions) here -- I think they apply equally to both pseudos, but I've only made them once.

lilles commented on css-pseudo-4/Overview.bs:
```diff
@@ -1634,6 +1634,32 @@ Styling form control pickers: the ''::picker()'' pseudo-element</h3>
+<h3 id="check">
+Option checkmark icon: the ''::check'' pseudo-element</h3>
+
+  The '::check'' pseudo-element only matches when the <a>originating
+  element</a> is a <{option}> element.
```
As I commented in the chromium prototype CL, should this instead say that the pseudo-element is only generated for option elements? That's an observable difference for `getComputedStyle(div, '::check')`.

josepharhar replied:
I changed it to say generated instead of matches

josepharhar commented:
> I mostly copied the definitions from before/after. If before/after aren't actually part-like, then should I copy the text from ::backdrop that looks like this instead of saying these pseudos are part like?
> 
> > The ::backdrop pseudo-element is a tree-abiding pseudo-element. It accepts all properties. It inherits from its originating element.

It sounds like this is ok, so I copied the text from ::backdrop to make it not part-like anymore

fantasai reviewed (changes requested):
Please rebase this over the css-forms-1 draft. :)

josepharhar commented:
> Please rebase this over the css-forms-1 draft. :)

done

tabatkins commented:
Per the resolutions from this week's call, these should be renamed to `::checkmark` and `::picker-icon`.

Also, I thought from the discussion in the call that `::checkmark` was meant to reflect the check in a checkbox/radio as well, but this currently defines it to only exist on option elements. I presume this is just old text that doesn't reflect more recent discussion?

josepharhar commented:
> Also, I thought from the discussion in the call that ::checkmark was meant to reflect the check in a checkbox/radio as well, but this currently defines it to only exist on option elements. I presume this is just old text that doesn't reflect more recent discussion? Similarly, I was under the impression that ::picker-icon would apply to several input types as well, not just select.

Yes people were talking about where else these pseudos could be used in order to get better names so they could be used for more things than select and options.

I'm not exactly sure what all the cases these pseudos will be used in yet because I only have a full definition/implementation of base appearance for select and not for all the other form controls.

I changed the definition for ::picker-icon which elements to match to be more vague/generic to support other future elements.

I changed the definition of ::checkmark to apply to anything which supports :checked and has base appearance.

tabatkins commented:
Ah, k, if that talk about other usage sites was somewhat theoretical, and they'll be handled in a different PR (if at all), then that's fine. I was just a little confused.
